### PR TITLE
Feat/improvements

### PR DESCRIPTION
**Pull request by Avantol13**

### CODE REVIEW

**Comment by PlanXCyborg** [comment]
This PR contains code that is not formatted correctly according to [`black`](https://github.com/ambv/black). Run `black` on your code before merging.

<details>
<summary>Expand the full diff to see formatting changes</summary>

```diff
--- gen3/index.py
+++ blackened
@@ -1,14 +1,17 @@
 import json
 import requests
 from urllib.parse import urljoin
 
+
 class Gen3Error(Exception):
     pass
 
+
 class Gen3UserError(Gen3Error):
     pass
+
 
 def handle_error(resp):
     if 400 <= resp.status_code < 600:
         try:
             json = resp.json()
@@ -16,15 +19,17 @@
         except KeyError:
             pass
         finally:
             resp.raise_for_status()
 
+
 def json_dumps(data):
-    return ({k: v for (k, v) in data.items() if v is not None})
+    return {k: v for (k, v) in data.items() if v is not None}
+
 
 class record_schema:
-    ''' A class to contain a the record schema
+    """ A class to contain a the record schema
 
         Schema:
             hashes (dict): {hash type: hash value,}
                 eg ``hashes={'md5': ab167e49d25b488939b1ede42752458b'}``
             size (int): file size metadata associated with a given uuid
@@ -35,57 +40,64 @@
             file_name (str): name of the file associated with a given UUID
             metadata (dict): additional key value metadata for this entry
             urls_metadata (dict): metadata attached to each url
             baseid (str): optional baseid to group with previous entries versions
             version (str): entry version string
-    '''
+    """
 
     def __init__(self, rdata):
-        '''
+        """
             rdata is a dictionary of the record data
-        '''
-        if all (k in rdata for k in ['hashes', 'size']) \
-                  and isinstance(rdata['hashes'], dict):
-            self.acl = rdata.get('acl', None)
-            self.authz = rdata.get('authz', None)
-            self.baseid = rdata.get('baseid', None)
-            self.did = rdata.get('did', None)
-            self.file_name = rdata.get('file_name', None)
-            self.form = rdata.get('form', 'object')
-            self.hashes = rdata['hashes']
-            self.metadata = rdata.get('metadata',None)
-            self.size = rdata['size']
-            self.uploader = rdata.get('uploader', None)
-            self.urls = rdata.get('urls', [])
-            self.urls_metadata = rdata.get('urls_metadata', None)
-            self.version = rdata.get('version', None)
+        """
+        if all(k in rdata for k in ["hashes", "size"]) and isinstance(
+            rdata["hashes"], dict
+        ):
+            self.acl = rdata.get("acl", None)
+            self.authz = rdata.get("authz", None)
+            self.baseid = rdata.get("baseid", None)
+            self.did = rdata.get("did", None)
+            self.file_name = rdata.get("file_name", None)
+            self.form = rdata.get("form", "object")
+            self.hashes = rdata["hashes"]
+            self.metadata = rdata.get("metadata", None)
+            self.size = rdata["size"]
+            self.uploader = rdata.get("uploader", None)
+            self.urls = rdata.get("urls", [])
+            self.urls_metadata = rdata.get("urls_metadata", None)
+            self.version = rdata.get("version", None)
         else:
-            raise Gen3UserError("record_schema must be initialized with a" \
-                                    " dictionary containing at least 'hashes'(dict)" \
-                                " and 'size'(int)")
+            raise Gen3UserError(
+                "record_schema must be initialized with a"
+                " dictionary containing at least 'hashes'(dict)"
+                " and 'size'(int)"
+            )
 
     def to_json(self):
-        '''
+        """
             Make the schema into json
-        '''
-        return json_dumps({
-            'acl': self.acl,
-            'authz': self.authz,
-            'baseid': self.baseid,
-            'did': self.did,
-            'file_name': self.file_name,
-            'form': self.form,
-            'hashes': self.hashes,
-            'metadata': self.metadata,
-            'size': self.size,
-            'uploader': self.uploader,
-            'urls': self.urls,
-            'urls_metadata': self.metadata,
-            'version': self.version
-        })
+        """
+        return json_dumps(
+            {
+                "acl": self.acl,
+                "authz": self.authz,
+                "baseid": self.baseid,
+                "did": self.did,
+                "file_name": self.file_name,
+                "form": self.form,
+                "hashes": self.hashes,
+                "metadata": self.metadata,
+                "size": self.size,
+                "uploader": self.uploader,
+                "urls": self.urls,
+                "urls_metadata": self.metadata,
+                "version": self.version,
+            }
+        )
+
 
 # ------------------------------------------------------------------------------
+
 
 class Gen3Index:
     """
 
     A class for interacting with the Gen3 Index services.
@@ -106,107 +118,105 @@
 
     """
 
     def __init__(self, endpoint, auth_provider):
         self._auth = auth_provider
-        self._endpoint = (endpoint + '/index/')
+        self._endpoint = endpoint + "/index/"
 
     def make_url(self, *path):
         # Creates a full url
-        return urljoin(self._endpoint, '/'.join(path))
-
-### Get Requests
+        return urljoin(self._endpoint, "/".join(path))
+
+    ### Get Requests
 
     def get_status(self):
-        ''' Return if indexd is healthy or not
-
-        '''
-        response = self._get('_status')
+        """ Return if indexd is healthy or not
+
+        """
+        response = self._get("_status")
         if response.status_code == 200:
             return response, "Indexd is Healthy!"
         else:
             return response, "Something is Wrong!"
 
     def get_version(self):
-        ''' Return the version of indexd
-
-        '''
-        return self._get('_version')
+        """ Return the version of indexd
+
+        """
+        return self._get("_version")
 
     def get_stats(self):
-        ''' Return basic info about the records in indexd
-
-        '''
-        return self._get('_stats')
+        """ Return basic info about the records in indexd
+
+        """
+        return self._get("_stats")
 
     def global_get(self, guid):
-        ''' Get the metadata associated with the given id, alias, or
+        """ Get the metadata associated with the given id, alias, or
             distributed identifier
 
         Args:
              guid: string
                  - record id
 
-        '''
-        return self._get(f'{guid}')
+        """
+        return self._get(f"{guid}")
 
     def get_urls(self, **params):
-        ''' Get a list of urls that match query params
+        """ Get a list of urls that match query params
 
         Args:
             size: integer
                 - object size
             hash: string
                 - hashes specified as algorithm:value
             ids: string
                 - comma delimited ids
 
-        '''
-        return self._get('index/urls', params=params) 
-
+        """
+        return self._get("index/urls", params=params)
 
     def get_index(self):
-        ''' Get a list of all records
-
-        '''
-        return self._get('index/') 
+        """ Get a list of all records
+
+        """
+        return self._get("index/")
 
     def get_record(self, guid):
-        ''' Get the metadata associated with a given id
-
-        '''
-        return self._get(f'/index/{guid}')
-
-    def get_latestversion(self, guid, has_version='true'):
-        ''' Get the metadata of the latest index record version associated
+        """ Get the metadata associated with a given id
+
+        """
+        return self._get(f"/index/{guid}")
+
+    def get_latestversion(self, guid, has_version="true"):
+        """ Get the metadata of the latest index record version associated
             with the given id
 
         Args:
             guid: string
                 - record id
             has_version: optional boolean string
                 - filter by latest doc that has the version value populated
 
-        '''
-        return self._get(f'/index/{guid}/latest', params=has_version)
+        """
+        return self._get(f"/index/{guid}/latest", params=has_version)
 
     def get_versions(self, guid):
-        ''' Get the metadata of index record version associated with the
+        """ Get the metadata of index record version associated with the
             given id
 
         Args:
             guid: string
                 - record id
 
-        '''
-        return self._get(f'/index/{guid}/versions')
-
-
-### Post Requests
+        """
+        return self._get(f"/index/{guid}/versions")
+
+    ### Post Requests
 
     def add_record(self, body):
-        ''' Create a new record and add it to the index
+        """ Create a new record and add it to the index
 
         Args:
             body: json/dictionary format
                 - Metadata object that needs to be added to the store.
                   Providing size and at least one hash is necessary and
@@ -216,33 +226,29 @@
             { 
                 "did": "string",
                 "baseid": "string",
                 "rev": "string"
             }
-        '''
+        """
         rec_data = record_schema(body)
-        return self._post(
-            'index',
-            auth=self._auth,
-            json=rec_data.to_json())
+        return self._post("index", auth=self._auth, json=rec_data.to_json())
 
     def create_blank(self, json):
-        ''' Create a blank record
+        """ Create a blank record
 
         Args:
             json - json in the format:
                     {
                         'uploader': type(string)
                         'file_name': type(string) (optional*)
                       }
 
-        '''
-        return self._post('index/blank', auth=self._auth, json=json)
-
+        """
+        return self._post("index/blank", auth=self._auth, json=json)
 
     def add_new_version(self, guid, body):
-        ''' Add new version for the document associated to the provided uuid
+        """ Add new version for the document associated to the provided uuid
 
             - Since data content is immutable, when you want to change the
               size or hash, a new index document with a new uuid needs to be
               created as its new version. That uuid is returned in the did
               field of the response. The old index document is not deleted.
@@ -254,52 +260,47 @@
                 - Metadata object that needs to be added to the store.
                   Providing size and at least one hash is necessary and
                   sufficient. Note: it is a good idea to add a version 
                   number of description
 
-        '''
+        """
         rec_data = record_schema(body)
-        return self._post(
-            f'index/{guid}',
-            auth=self._auth,
-            json=rec_data.to_json())
+        return self._post(f"index/{guid}", auth=self._auth, json=rec_data.to_json())
 
     def get_record_bulk(self, body):
-        ''' Get a list of documents given a list of dids
+        """ Get a list of documents given a list of dids
 
         Args:
             body: list
                  - a list of record ids
 
-        '''
-        return self._post('bulk/documents', auth=self._auth, json=body) 
-
-### Put Requests
+        """
+        return self._post("bulk/documents", auth=self._auth, json=body)
+
+    ### Put Requests
 
     def update_blank(self, guid, rev, body):
-        ''' Update only hashes and size for a blank index
+        """ Update only hashes and size for a blank index
 
         Args:
              guid: string
                   - record id
              rev: string
                 - data revision - simple consistency mechanism
              body: json/dictionary format
                  - index record that needs to be updated.
                    *Size and hashes are required fields.
 
-        '''
+        """
         rec_data = record_schema(body)
-        p = {'rev': rev}
+        p = {"rev": rev}
         return self._put(
-                f'index/blank/{guid}',
-                params=p,
-                auth=self._auth,
-                json=rec_data.to_json()) 
+            f"index/blank/{guid}", params=p, auth=self._auth, json=rec_data.to_json()
+        )
 
     def update_record(self, guid, rev, body):
-        ''' Update an existing entry in the index
+        """ Update an existing entry in the index
 
         Args:
              guid: string
                  - record id
             rev: string
@@ -324,34 +325,30 @@
                 "did": "string",
                 "baseid": "string",
                 "rev": "string"
             }
 
-        '''
-        p = {'rev': rev}
-        return self._put(
-                f'index/{guid}',
-                params=p,
-                auth=self._auth,
-                json=body)
-
-### Delete Requests
+        """
+        p = {"rev": rev}
+        return self._put(f"index/{guid}", params=p, auth=self._auth, json=body)
+
+    ### Delete Requests
 
     def delete_record(self, guid, rev):
-        ''' Delete an entry from the index
+        """ Delete an entry from the index
 
         Args:
             guid: string
                  - record id
             rev: string
                  - data revision - simple consistency mechanism
 
-        '''
-        p = {'rev': rev}
-        return self._delete(f'index/{guid}', auth=self._auth, params=p)
-
-### Useful Helper functions for http requests
+        """
+        p = {"rev": rev}
+        return self._delete(f"index/{guid}", auth=self._auth, params=p)
+
+    ### Useful Helper functions for http requests
     def _get(self, *path, **kwargs):
         resp = requests.get(self.make_url(*path), **kwargs)
         handle_error(resp)
         return resp
 

--- gen3/submission.py
+++ blackened
@@ -219,12 +219,12 @@
         """
         api_url = "{}/api/v0/submission/{}/{}".format(self._endpoint, program, project)
         output = requests.delete(api_url, auth=self._auth_provider).text
         return output
 
-#**********************************************************************************************
-# 7/16/2019 - Michael Martinson
+    # **********************************************************************************************
+    # 7/16/2019 - Michael Martinson
 
     def view_programs(self):
         """List registered programs
 
         """
@@ -294,11 +294,11 @@
         """
         api_url = f"{self._endpoint}/api/v0/submission/{program}/{project}/open"
         output = requests.put(api_url, auth=self._auth_provider).text
         return output
 
-#**********************************************************************************************
+    # **********************************************************************************************
 
     def create_program(self, json):
         """Create a program.
         Args:
             json (object): The json of the program to create
@@ -398,12 +398,12 @@
         # Read the file in as a pandas DataFrame
         f = os.path.basename(filename)
         if f.lower().endswith(".csv"):
             df = pd.read_csv(filename, header=0, sep=",", dtype=str).fillna("")
         elif f.lower().endswith(".xlsx"):
-            #xl = pd.ExcelFile(filename, dtype=str)  # load excel file
-            # This ^ was previous code but ExcelFile() does not have an argument dtype 
+            # xl = pd.ExcelFile(filename, dtype=str)  # load excel file
+            # This ^ was previous code but ExcelFile() does not have an argument dtype
             # so I removed it and it works
             xl = pd.ExcelFile(filename)  # load excel file
             sheet = xl.sheet_names[0]  # sheetname
             df = xl.parse(sheet)  # save sheet as dataframe
             converters = {
@@ -412,11 +412,13 @@
             df = pd.read_excel(filename, converters=converters).fillna("")  # remove nan
         elif filename.lower().endswith((".tsv", ".txt")):
             df = pd.read_csv(filename, header=0, sep="\t", dtype=str).fillna("")
         else:
             raise Gen3UserError("Please upload a file in CSV, TSV, or XLSX format.")
-        df.rename(columns = {c: c.lstrip('*') for c in df.columns}, inplace = True) # remove any leading asterisks in the DataFrame column names
+        df.rename(
+            columns={c: c.lstrip("*") for c in df.columns}, inplace=True
+        )  # remove any leading asterisks in the DataFrame column names
 
         # Check uniqueness of submitter_ids:
         if len(list(df.submitter_id)) != len(list(df.submitter_id.unique())):
             raise Gen3Error(
                 "Warning: file contains duplicate submitter_ids. \nNote: submitter_ids must be unique within a node!"

--- setup.py
+++ blackened
@@ -5,11 +5,11 @@
 def get_version():
     try:
         tag = check_output(
             ["git", "describe", "--tags", "--abbrev=0", "--match=[0-9]*"]
         )
-        return tag.decode('utf-8').strip("\n")
+        return tag.decode("utf-8").strip("\n")
     except Exception:
         # if somehow you get the repo not from git,
         # hardcode default major version
         return "0.1.0"
 

--- tests/indexd_test.py
+++ blackened
@@ -1,225 +1,242 @@
 import pytest
 from requests import HTTPError
 
 # helper functions -------------------------------------------------
 def get_rec(indexd, guid):
-    # testing get_record 
+    # testing get_record
     return indexd.get_record(guid).json()
+
 
 def no_record(indexd, guid):
     try:
         get_rec(indexd, guid)
 
         return False
     except HTTPError as error:
-        if 'no record found' in str(error):
+        if "no record found" in str(error):
             return True
         else:
             raise "Something is wrong!"
 
+
 def same_recs(rec1, rec2):
-    rec1.pop('rev', None)
-    rec1.pop('version', None)
-    rec2.pop('rev', None)
-    rec2.pop('version', None)
-    print(rec1, '\n', rec2)
+    rec1.pop("rev", None)
+    rec1.pop("version", None)
+    rec2.pop("rev", None)
+    rec2.pop("version", None)
+    print(rec1, "\n", rec2)
     if rec1 == rec2:
         return True
     else:
         return False
+
+
 # tests ------------------------------------------------------------
 
+
 def testsystem(indexd):
-    ''' Test that indexd is healthy 
-    '''
+    """ Test that indexd is healthy 
+    """
     assert indexd.get_status()[0].status_code == 200
     assert indexd.get_version().status_code == 200
     assert indexd.get_stats().status_code == 200
     assert indexd.get_index().status_code == 200
-#-------------------------------------------------------------------
+
+
+# -------------------------------------------------------------------
+
 
 def testbulk(indexd):
-    ''' Test get_record_bulk
-    '''
-     # put a new record in the index
+    """ Test get_record_bulk
+    """
+    # put a new record in the index
     rec1 = indexd.add_record(
-            {
-                'authz': ['/programs/prog1/projects/proj1'],
-                'hashes': {'md5':'374c12456782738abcfe387492837483'},
-                'size': 0
-                }
-            ).json()
+        {
+            "authz": ["/programs/prog1/projects/proj1"],
+            "hashes": {"md5": "374c12456782738abcfe387492837483"},
+            "size": 0,
+        }
+    ).json()
     # put a new record in the index
     rec2 = indexd.add_record(
-            {
-                'authz': ['/programs/prog1/projects/proj1'],
-                'hashes': {'md5':'adbc12447582738abcfe387492837483'},
-                'size': 0
-                }
-            ).json()
+        {
+            "authz": ["/programs/prog1/projects/proj1"],
+            "hashes": {"md5": "adbc12447582738abcfe387492837483"},
+            "size": 0,
+        }
+    ).json()
     # put a new record in the index
     rec3 = indexd.add_record(
-            {
-                'authz': ['/programs/prog1/projects/proj1'],
-                'hashes': {'md5':'adbc82746782738abcfe387492837483'},
-                'size': 0
-                }
-            ).json()
-
-    recs = indexd.get_record_bulk(
-            [
-                rec1['did'],
-                rec2['did'],
-                rec3['did']
-                ]
-            )
-
-    dids = [rec1['did']] + [rec2['did']] + [rec3['did']]
+        {
+            "authz": ["/programs/prog1/projects/proj1"],
+            "hashes": {"md5": "adbc82746782738abcfe387492837483"},
+            "size": 0,
+        }
+    ).json()
+
+    recs = indexd.get_record_bulk([rec1["did"], rec2["did"], rec3["did"]])
+
+    dids = [rec1["did"]] + [rec2["did"]] + [rec3["did"]]
     v = True
     for rec in recs.json():
-        if rec['did'] not in dids:
+        if rec["did"] not in dids:
             v = False
 
     assert v
-#-------------------------------------------------------------------
+
+
+# -------------------------------------------------------------------
+
 
 def testnewrecord(indexd):
-    ''' Test the creation, update, and deletion a record
+    """ Test the creation, update, and deletion a record
 
         index.py functions tested:
             add_record 
             global_get
             get_record
             update_record
             delete_record
-    '''
+    """
 
     # put a new record in the index
     newrec = indexd.add_record(
-            {
-                'authz': ['/programs/prog1/projects/proj1'],
-                'hashes': {'md5':'adbc12456782738abcfe387492837483'},
-                'size':0
-                }
-            ).json()
+        {
+            "authz": ["/programs/prog1/projects/proj1"],
+            "hashes": {"md5": "adbc12456782738abcfe387492837483"},
+            "size": 0,
+        }
+    ).json()
 
     # testing global get
-    checkrec = indexd.global_get(newrec['did']).json()
-    assert newrec['did'] == checkrec['did'] and \
-           newrec['baseid'] == checkrec['baseid'] and \
-           newrec['rev'] == checkrec['rev']
+    checkrec = indexd.global_get(newrec["did"]).json()
+    assert (
+        newrec["did"] == checkrec["did"]
+        and newrec["baseid"] == checkrec["baseid"]
+        and newrec["rev"] == checkrec["rev"]
+    )
 
     # update the record
     updated = indexd.update_record(
-            newrec['did'], 
-            newrec['rev'],
-            {'acl':['prog1', 'proj1'],
-                'file_name':'fakefilename',
-                'uploader':'mjmartinson'}).json()
-
-    updatedrec = get_rec(indexd, updated['did'])
-    # Note: I am not sure why the program and project are flipped!! 
-    assert updatedrec['acl'] == ['proj1', 'prog1'] and \
-           updatedrec['file_name'] == 'fakefilename' and \
-           updatedrec['uploader'] == 'mjmartinson' and \
-           updatedrec['did'] == checkrec['did'] and \
-           updatedrec['rev'] != checkrec['rev']
-
-    #delete the record
-    response = indexd.delete_record(updatedrec['did'], updatedrec['rev']) 
-    assert response.status_code == 200 and \
-            no_record(indexd, updatedrec['did']) == True
-#-------------------------------------------------------------------
+        newrec["did"],
+        newrec["rev"],
+        {
+            "acl": ["prog1", "proj1"],
+            "file_name": "fakefilename",
+            "uploader": "mjmartinson",
+        },
+    ).json()
+
+    updatedrec = get_rec(indexd, updated["did"])
+    # Note: I am not sure why the program and project are flipped!!
+    assert (
+        updatedrec["acl"] == ["proj1", "prog1"]
+        and updatedrec["file_name"] == "fakefilename"
+        and updatedrec["uploader"] == "mjmartinson"
+        and updatedrec["did"] == checkrec["did"]
+        and updatedrec["rev"] != checkrec["rev"]
+    )
+
+    # delete the record
+    response = indexd.delete_record(updatedrec["did"], updatedrec["rev"])
+    assert response.status_code == 200 and no_record(indexd, updatedrec["did"]) == True
+
+
+# -------------------------------------------------------------------
 @pytest.mark.skip
 def testversions(indexd):
-    ''' Test creation of a record and a new version of it
+    """ Test creation of a record and a new version of it
 
         index.py functions tested:
             add_record
             add_new_version
             get_versions
             get_latestversion
-    '''
+    """
 
     # put a new record in the index
     newrec = indexd.add_record(
-            {
-                'acl': ['prog1', 'proj1'],
-                'authz': ['/programs/prog1/projects/proj1'],
-                'hashes': {'md5':'437283456782738abcfe387492837483'},
-                'size': 0,
-                'version': '1',
-                'uploader': 'mjmartinson'
-                }
-            ).json()
+        {
+            "acl": ["prog1", "proj1"],
+            "authz": ["/programs/prog1/projects/proj1"],
+            "hashes": {"md5": "437283456782738abcfe387492837483"},
+            "size": 0,
+            "version": "1",
+            "uploader": "mjmartinson",
+        }
+    ).json()
 
     # update the record
     newversion = indexd.add_new_version(
-            newrec['did'], 
-            {
-                'acl': ['prog1', 'proj1'],
-                'authz': ['/programs/prog1/projects/proj1'],
-                'hashes': {'md5': '437283456782738abcfe387492837483'},
-                'size': 1,
-                'version': '2',
-                'uploader': 'mjmartinson'
-                }).json()
-
-    return get_rec(indexd, newrec['did'])
-    return get_rec(indexd, newversion['did'])
+        newrec["did"],
+        {
+            "acl": ["prog1", "proj1"],
+            "authz": ["/programs/prog1/projects/proj1"],
+            "hashes": {"md5": "437283456782738abcfe387492837483"},
+            "size": 1,
+            "version": "2",
+            "uploader": "mjmartinson",
+        },
+    ).json()
+
+    return get_rec(indexd, newrec["did"])
+    return get_rec(indexd, newversion["did"])
 
     print(newrec)
     print(newversion)
 
-#   These functions do not recognize the records for some reason!
-    versions = indexd.get_versions(newversion['did'])
-    latestversion = indexd.get_latestversion(newrec['did'], 'false')
-#-------------------------------------------------------------------
+    #   These functions do not recognize the records for some reason!
+    versions = indexd.get_versions(newversion["did"])
+    latestversion = indexd.get_latestversion(newrec["did"], "false")
+
+
+# -------------------------------------------------------------------
 
 # the endpoint /blank is having some sort of authorization problem
 # it asks for username and password even when given auth file
-@pytest.mark.skip 
+@pytest.mark.skip
 def testblank(indexd):
-    ''' Test create and update blank record
-     '''
-    newblank = indexd.create_blank(
-            {
-                'uploader': 'mjmartinson'
-                }
-            )
-
-    checkrec = get_rec(indexd, newblank['did'])
-    assert newblank['did'] == checkrec['did'] and \
-            newblank['baseid'] == checkrec['baseid'] and \
-            newblank['rev'] == checkrec['rev']
+    """ Test create and update blank record
+     """
+    newblank = indexd.create_blank({"uploader": "mjmartinson"})
+
+    checkrec = get_rec(indexd, newblank["did"])
+    assert (
+        newblank["did"] == checkrec["did"]
+        and newblank["baseid"] == checkrec["baseid"]
+        and newblank["rev"] == checkrec["rev"]
+    )
 
     # update the record
     updated = indexd.update_blank(
-            newblank['did'], 
-            newblank['rev'],
-            {
-                'size': 1,
-                'hashes': {'md5': '4372834515237483626e387492837483'}
-                }).json()
-
-    updatedblank = get_rec(indexd, updated['did'])
-    assert updatedblank['did'] == checkrec['did'] and \
-            updatedblank['size'] == 1 and \
-            updatedblank['did'] == {'md5': '4372834515237483626e387492837483'} \
-            and updatedblank['rev'] != checkrec['rev']
-
-    #delete the record
-    response = indexd.delete_record(updatedblank['did'], updatedblank['rev']) 
-    assert response.status_code == 200 and \
-            no_record(indexd, updatedblank['did']) == True
-#-------------------------------------------------------------------
-
-''' Functions that Do not work right now 7/25
+        newblank["did"],
+        newblank["rev"],
+        {"size": 1, "hashes": {"md5": "4372834515237483626e387492837483"}},
+    ).json()
+
+    updatedblank = get_rec(indexd, updated["did"])
+    assert (
+        updatedblank["did"] == checkrec["did"]
+        and updatedblank["size"] == 1
+        and updatedblank["did"] == {"md5": "4372834515237483626e387492837483"}
+        and updatedblank["rev"] != checkrec["rev"]
+    )
+
+    # delete the record
+    response = indexd.delete_record(updatedblank["did"], updatedblank["rev"])
+    assert (
+        response.status_code == 200 and no_record(indexd, updatedblank["did"]) == True
+    )
+
+
+# -------------------------------------------------------------------
+
+""" Functions that Do not work right now 7/25
 
     get_urls - do not understand input parameters
     get_latestversion
     get_versions
     create_blank
     update_blank
-'''
-
+"""
+
```
</details></br>

This formatting comment was generated automatically by a script in [uc-cdis/wool](https://github.com/uc-cdis/wool).

File: tests/test_import.py [hunk 1]
```diff
@@ -1,23 +0,0 @@
-def test_import():
-
-    import gen3
```
Comment (Avantol13) [inline tests/test_import.py]: why did you remove all these import tests?

File: gen3/submission.py [hunk 2]
```diff
@@ -323,7 +407,10 @@ def submit_file(self, project_id, filename, chunk_size=30, row_offset=0):
         if f.lower().endswith(".csv"):
             df = pd.read_csv(filename, header=0, sep=",", dtype=str).fillna("")
         elif f.lower().endswith(".xlsx"):
-            xl = pd.ExcelFile(filename, dtype=str)  # load excel file
+            # xl = pd.ExcelFile(filename, dtype=str)  # load excel file
```
Comment (Avantol13) [inline gen3/submission.py]: you can just remove these 3 lines / comments and leave the fixed line

File: gen3/index.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+        self.client = client.IndexClient(endpoint, auth=auth_provider)
+
+    ### Get Requests
+
+    def get_status(self):
+        """ Return if indexd is healthy or not
+
+        """
+        response = self.client._get("_status")
+        if response.status_code == 200:
```
Comment (Avantol13) [inline gen3/index.py]: I would just return the response without checking the code. e.g. `return self.client._get("_status)` to be more like the other apis here

File: gen3/index.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+            return response, "Indexd is Healthy!"
+        else:
+            return response, "Something is Wrong!"
+
+    def get_version(self):
+        """ Return the version of indexd
+
+        """
+        return self.client._get("_version").json()
```
Comment (Avantol13) [inline gen3/index.py]: we may want to also raise exceptions if there's an error response. note that .json does _not_ do that: https://2.python-requests.org//en/latest/user/quickstart/#json-response-content

Comment (Avantol13) [reply gen3/index.py]: so we could use the pattern:
```
response = self.client._get("_version")
response.raise_for_status()
return response.json()
```

File: tests/test_import.py (hunk 1 above)
Comment (mjmartinson) [reply tests/test_import.py]: That was an accident.

File: gen3/index.py [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+
+    def get_stats(self):
+        """ Return basic info about the records in indexd
+
+        """
+        response = self.client._get("_stats")
+        response.raise_for_status()
+        return response.json()
+
+    def get_index(self):
```
Comment (Avantol13) [inline gen3/index.py]: lets change these function name to be more user friendly. LIke this could be `get_all_records`

File: gen3/index.py [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+        """ Get a list of all records
+
+        """
+        response = self.client._get("index/")
+        response.raise_for_status()
+        return response.json()
+
+    def global_get(self, guid, no_dist=False):
```
Comment (Avantol13) [inline gen3/index.py]: let's change this to just `get` since it should be the default

File: gen3/index.py [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+        as ``{'hashes': {'md5': '...'}, 'size': '...', 'metadata': {'file_state': '...'}}``.
+
+            - need to include all the hashes in the request
+            - index client like signpost or indexd will need to handle the
+              query param `'hash': 'hash_type:hash'`
+        """
+        rec = self.client.get_with_params(params)
+        return rec.to_json()
+
+    def get_latestversion(self, guid, has_version=False):
```
Comment (Avantol13) [inline gen3/index.py]: can you change to `get_latest_version`?

File: gen3/index.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+        """
+        response = self.client._get(f"/index/{guid}/versions")
+        response.raise_for_status()
+        versions = response.json()
+
+        return [r for _, r in versions.items()]
+
+    ### Post Requests
+
+    def add_record(
```
Comment (Avantol13) [inline gen3/index.py]: it's minor, but can this be `create_record`?

File: gen3/index.py [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+            auth=self.client.auth,
+        )
+        response.raise_for_status()
+        rec = response.json()
+
+        if rec and "did" in rec:
+            return self.get_record(rec["did"])
+        return None
+
+    def get_record_bulk(self, dids):
```
Comment (Avantol13) [inline gen3/index.py]: I think just `get_records` would be a better name

File: gen3/index.py [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+
+    """
+
+    def __init__(self, endpoint, auth_provider):
+        self.client = client.IndexClient(endpoint, auth=auth_provider)
+
+    ### Get Requests
+
+    def get_status(self):
+        """ Return if indexd is healthy or not
```
Comment (Avantol13) [inline gen3/index.py]: for our docstrings our style is typically to put this on the next line, did you run `black` against this? Did it do that?

Comment (mjmartinson) [reply gen3/index.py]: That is just how I did it. I ran black but it did not change it. I will change them.

File: test-requirements.txt [hunk 11]
```diff
@@ -1 +1,14 @@
 pytest>=3.2.3
+mock
+
+# these are for indexd
+authutils==3.1.0
+gen3rbac==0.1.2
+
+-e git+https://github.com/uc-cdis/indexclient.git@1.6.0#egg=indexclient
```
Comment (Avantol13) [inline test-requirements.txt]: some of these need to be in the requirements, not test-requirements. any library used in the actual SDK (not tests) so go in requirements. so indexclient, cdis-python-utils, maybe cdispyutils and cdislogging

File: test-requirements.txt [hunk 12]
```diff
@@ -1 +1,14 @@
...
+-e git+https://github.com/uc-cdis/cdisutils-test.git@1.0.0#egg=cdisutilstest
+-e git+https://github.com/uc-cdis/cdis-python-utils.git@0.1.7#egg=cdispyutils
+-e git+https://github.com/uc-cdis/indexd.git@2.0.0#egg=indexd
+-e git+https://github.com/uc-cdis/doiclient.git@1.0.0#egg=doiclient
+-e git+https://github.com/uc-cdis/dosclient.git@1.0.0#egg=dosclient
+-e git+https://github.com/uc-cdis/cdislogging.git@0.0.2#egg=cdislogginge git+https://github.com/uc-cdis/cdis-python-utils.git@0.1.7#egg=cdispyutils
```
Comment (Avantol13) [inline test-requirements.txt]:
```suggestion
-e git+https://github.com/uc-cdis/cdislogging.git@0.0.2#egg=cdislogging
git+https://github.com/uc-cdis/cdis-python-utils.git@0.1.7#egg=cdispyutils
```

File: tests/test_import.py [hunk 13]
```diff
@@ -1,11 +1,19 @@
...
 
-def test_indexclient_import():
+def test_cdisutilstest():
+
+    import cdisutilstest
+
+
+def test_indexclient():
 
-    from gen3.indexclient.client import IndexClient
```
Comment (Avantol13) [inline tests/test_import.py]: this should still work after we did the `cp -r` fix, was it failing?

Comment (mjmartinson) [reply tests/test_import.py]: Yes I think so

File: gen3/submission.py [hunk 14]
```diff
@@ -49,43 +49,179 @@ def __export_file(self, filename, output):
         outfile.close
         print("\nOutput written to file: " + filename)
 
-    def query(self, query_txt, variables=None, max_tries=1):
-        """Execute a GraphQL query against a data commons.
+    ### Program functions
+
+    def view_programs(self):
```
Comment (Avantol13) [inline gen3/submission.py]: can you change to `get_programs` and any other instance of `view` to `get` for consistency with other function names?

File: gen3/submission.py [hunk 15]
```diff
@@ -49,43 +49,179 @@ def __export_file(self, filename, output):
         outfile.close
         print("\nOutput written to file: " + filename)
 
-    def query(self, query_txt, variables=None, max_tries=1):
```
Comment (Avantol13) [inline gen3/submission.py]: the diff here is being weird, I think (and can you confirm) that you didn't change any of the existing functions? Seems like you just added new ones above these and github isn't lining things up correctly

Comment (mjmartinson) [reply gen3/submission.py]: Yes, those are all the same.

File: tests/unitindexd_test.py [hunk 16]
```diff
@@ -0,0 +1,234 @@
+import pytest
+from requests import HTTPError
+
+
+def get_rec(gen3_index, guid):
+    # testing get_record
+    return gen3_index.get_record(guid)
+
+
+def testsystem(gen3_index):
```
Comment (Avantol13) [inline tests/unitindexd_test.py]: can you separate these words with `_` and for future tests?

Comment (mjmartinson) [reply tests/unitindexd_test.py]: will do.

File: gen3/index.py [hunk 17]
```diff
@@ -0,0 +1,420 @@
...
+        This generates the Gen3Index class pointed at the sandbox commons while
+        using the credentials.json downloaded from the commons profile page.
+
+        >>> endpoint = "https://nci-crdc-demo.datacommons.io"
+        ... auth = Gen3Auth(endpoint, refresh_file="credentials.json")
+        ... sub = Gen3Submission(endpoint, auth)
+
+    """
+
+    def __init__(self, endpoint, auth_provider):
```
Comment (Avantol13) [inline gen3/index.py]: you should allow not providing an auth_provider, e.g. default it to None. b/c some of the functions don't require access (like reading records)

Comment (mjmartinson) [reply gen3/index.py]: Should I change this in submission.py also?
